plugins/pam.lua

```diff
@@ -0,0 +1,264 @@
+-- Plugin for PAM attributes
+--
+-- Configuration:
+--   uid_offset: offset applied to all assigned uid's (default: 100000)
+--   gid_offset: offset applied to all assigned gid's (default: 100000)
+--   default_group: name of default group for users gid attribute (default: pam_users)
+--
+-- Creates the following:
+--   User Attributes: uidnumber, gidnumber
+--   Group Attributes: gidnumber
+--   Groups: pam_users (default)
+--
+-- Strives to maintain:
+--   Users: all have a unique uidnumber
+--   Users: all users by default have a gidnumber corresponding to the
+--          pam_users group created.
+--   Groups: all groups have a unique uidnumber
+--
+-- Group will only be created when a user needs the gidnumber for
+-- it, this is done to avoid it being created as group 1, before
+-- lldap_admin, in new setups.
+--
+-- TODO:
+--   3. max uid should be stored in a key-value store instead
+--
+
+-- keys used for this plugins key-value store
+local KVKEYS = {
+    NEXT_UID = "next-uid"
+}
+
+local get_config_gid_offset = function(context)
+    if context.configuration.gid_offset ~= nil then
+        return tonumber(context.configuration.gid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+
+local get_config_uid_offset = function(context)
+    if context.configuration.uid_offset ~= nil then
+        return tonumber(context.configuration.uid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+local get_config_default_group = function(context)
+    if context.configuration.default_group ~= nil then
+        return context.configuration.default_group
+    else
+        -- return default
+        return "pam_users"
+    end
+end
+
+local resolve_gid_default_group = function(context)
+    local group_name = get_config_default_group(context)
+    local groups = context.api:list_groups({ filter = "(cn=" .. group_name .. ")" })
+    local group_id = nil
+    if lldap.tables:empty(groups) then
+        lldap.log:debug("Creating group:" .. group_name)
+        local id, err = context.api:create_group({
+            display_name = group_name,
+            attributes = {}
+        })
+        if err ~= nil then
+            -- Error
+            error("Failed to create group: " .. group_name, 1)
+        end
+        lldap.log:debug("Created '" .. group_name .. "' group with id " .. tostring(id))
+        local group_gidnumber = get_config_gid_offset(context) + id
+        local res, err = context.api:update_group({
+            group_id = id,
+            insert_attributes = {
+                gidnumber = { int = group_gidnumber }
+            }
+        })
+        if err ~= nil then
+            error("Failed to set gidnumber attribute to '" .. group_name .. "' group", 1)
```

Comment:
Just jotting down my thoughts: We need to be careful/intentional about logging levels. Again, this is something I don't want plugin authors to have to think about, but I don't want to flood the logs just because you installed a plugin. That means that unless you enable debug logs or the plugins log an error/warning, you don't see them (apart from initialization). If they _do_ log an error, but you're not in debug logs, you might not see the whole stacktrace leading to the error, so the message has to be prepended with "plugin pam_plugin:" or similar.

Reply:
We could make lldap.log:{debug,info,error} prepend the plugin name automatically? The log object should maybe be moved to context then, so it has the information...

Just to be sure; the error function here is not for logging, it aborts the script.